sdks/java/io/hadoop-output-format/src/test/java/org/apache/beam/sdk/io/hadoop/outputformat/HadoopOutputFormatIOTest.java

```diff
@@ -0,0 +1,202 @@
+
+  @BeforeClass
+  public static void setUp() {
+    EmployeeOutputFormat.initWrittenOutput();
+    serConf = loadTestConfiguration(EmployeeOutputFormat.class, Text.class, Employee.class);
+  }
+
+  private static SerializableConfiguration loadTestConfiguration(
+      Class<?> outputFormatClassName, Class<?> keyClass, Class<?> valueClass) {
+    Configuration conf = new Configuration();
+    conf.setClass("mapreduce.job.outputformat.class", outputFormatClassName, OutputFormat.class);
+    conf.setClass("mapreduce.job.outputformat.key.class", keyClass, Object.class);
+    conf.setClass("mapreduce.job.outputformat.value.class", valueClass, Object.class);
+    return new SerializableConfiguration(conf);
+  }
+
+  @Test
+  public void testWriteBuildsCorrectly() {
+    HadoopOutputFormatIO.Write<Text, Employee> write =
+        HadoopOutputFormatIO.<Text, Employee>write().withConfiguration(serConf.get());
+
+    assertEquals(serConf.get(), write.getConfiguration().get());
+    assertEquals(EmployeeOutputFormat.class, write.getOutputFormatClass().getRawType());
+    assertEquals(Text.class, write.getOutputFormatKeyClass().getRawType());
+    assertEquals(Employee.class, write.getOutputFormatValueClass().getRawType());
+  }
+
+  /**
+   * This test validates {@link HadoopOutputFormatIO.Write Write} transform object creation fails
+   * with null configuration. {@link HadoopOutputFormatIO.Write#withConfiguration(Configuration)
+   * withConfiguration(Configuration)} method checks configuration is null and throws exception if
+   * it is null.
+   */
+  @Test
+  public void testWriteObjectCreationFailsIfConfigurationIsNull() {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Configuration can not be null");
+    HadoopOutputFormatIO.<Text, Employee>write().withConfiguration(null);
+  }
+
+  /**
+   * This test validates functionality of {@link
+   * HadoopOutputFormatIO.Write#withConfiguration(Configuration) withConfiguration(Configuration)}
+   * function when Hadoop OutputFormat class is not provided by the user in configuration.
+   */
+  @Test
+  public void testWriteValidationFailsMissingInputFormatInConf() {
+    Configuration configuration = new Configuration();
+    configuration.setClass(HadoopOutputFormatIO.OUTPUTFORMAT_KEY_CLASS, Text.class, Object.class);
+    configuration.setClass(
+        HadoopOutputFormatIO.OUTPUTFORMAT_VALUE_CLASS, Employee.class, Object.class);
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Configuration must contain \"mapreduce.job.outputformat.class\"");
+    HadoopOutputFormatIO.<Text, Employee>write().withConfiguration(configuration);
+  }
+
+  /**
+   * This test validates functionality of {@link
+   * HadoopOutputFormatIO.Write#withConfiguration(Configuration) withConfiguration(Configuration)}
+   * function when key class is not provided by the user in configuration.
+   */
+  @Test
+  public void testWriteValidationFailsMissingKeyClassInConf() {
```

Comment:
Ditto: it is equivalent of testing that `checkArgument`  used in validate method works

Reply:
We test here that in case of invalid configuration we throw proper exception with appropriate error message. So, it checks that `checkArgument` works correctly for this invalid configuration.

Reply:
yes but checkArgument was already tested in Google common

Reply:
Sorry, maybe I don't understand your point, but I don't test if `checkArgument` works by itself. I test that `checkArgument` was properly called during creation of pipeline and it throws appropriate error in case of invalid configuration.  For example, `BigtableIOTest` tests the same.

Reply:
It is just that it seems strange to me to test either that:
- checkArgument works: because already tested
- validateConfiguration is called because you explicitly call it yourself in withConfiguration
But nevermind.

Reply:
no objections of course, the more tests the better. It was just a remark